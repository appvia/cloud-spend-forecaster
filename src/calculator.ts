import { Component, ComponentOutput, ComputeNode, Interval } from '../module'

export function putComponentsIntoIntervals(components: Component[]) {
  return (interval: Interval): Interval => {
    return { ...interval, components }
  }
}

export function putNodesIntoIntervals(node: ComputeNode) {
  return (interval: Interval): Interval => {
    return { ...interval, nodes: node }
  }
}

export function calculateResourceNeededForRequestsInterval(interval: Interval): Interval {
  const intervalComponents = interval.components.map((component: ComponentOutput): ComponentOutput => {
    return {
      ...component,
      needCpuForRequests: interval.requests * component.requestToCpu,
      needMemoryForRequests: interval.requests * component.requestToMemory,
    }
  })
  return {
    ...interval,
    components: intervalComponents,
  }
}

export function calculateReplicasNeededForInterval(interval: Interval): Interval {
  const intervalComponents = interval.components.map((component: ComponentOutput): ComponentOutput => {
    const needCpuReplica = Math.ceil(component.needCpuForRequests / (component.limitCpu - component.baselineCpu))
    const needMemoryReplica = Math.ceil(
      component.needMemoryForRequests / (component.limitMemory - component.baselineMemory),
    )

    return {
      ...component,
      needCpuReplica,
      needMemoryReplica,
      needReplica: Math.max(needMemoryReplica, needCpuReplica, component.minReplica),
    }
  })
  return {
    ...interval,
    components: intervalComponents,
    needPods: intervalComponents.reduce((accumulator, component) => accumulator + component.needReplica, 0),
  }
}

export function calculateResourceNeededForInterval(interval: Interval): Interval {
  const intervalComponents = interval.components.map((component: ComponentOutput): ComponentOutput => {
    return {
      ...component,
      needCpu: component.needReplica * component.baselineCpu + component.needCpuForRequests,
      needMemory: component.needReplica * component.baselineMemory + component.needMemoryForRequests,
    }
  })
  return {
    ...interval,
    components: intervalComponents,
    needCpu: intervalComponents.reduce((accumulator, component) => accumulator + component.needCpu, 0),
    needMemory: intervalComponents.reduce((accumulator, component) => accumulator + component.needMemory, 0),
  }
}

export function calculateNodesNeededForInterval(interval: Interval): Interval {
  const needNodesByCpu = Math.ceil(interval.needCpu / interval.nodes.availableCpu)
  const needNodesByMemory = Math.ceil(interval.needMemory / interval.nodes.availableMemory)
  const needNodesByPods = Math.ceil(interval.needPods / interval.nodes.maxPods)

  return { ...interval, needNodesByCpu, needNodesByMemory, needNodesByPods }
}

export function calcCapacity(replica: number, limit: number, to_request: number, baseline: number): number {
  return Math.floor(replica * (limit * to_request - baseline))
}

export function withinRange(min: number, max: number, number: number): number {
  return Math.ceil(Math.max(min, Math.min(max, number)))
}

export function calculateDesiredResourceForInterval(interval: Interval): Interval {
  const intervalComponents = interval.components.map((component: ComponentOutput): ComponentOutput => {
    const desiredReplica = withinRange(
      component.minReplica,
      component.maxReplica,
      (100 / component.scalingThresholdCpu) * component.needCpuReplica,
    )
    const desiredCpu = desiredReplica * component.limitCpu
    const desiredMemory = desiredReplica * component.limitMemory
    return {
      ...component,
      desiredReplica,
      desiredCpu,
      desiredMemory,
    }
  })
  return {
    ...interval,
    components: intervalComponents,
    desiredPods: intervalComponents.reduce(
      (accumulator: number, component: ComponentOutput): number => accumulator + component.desiredReplica,
      0,
    ),
    desiredCpu: intervalComponents.reduce(
      (accumulator: number, component: ComponentOutput): number => accumulator + component.desiredCpu,
      0,
    ),
    desiredMemory: intervalComponents.reduce(
      (accumulator: number, component: ComponentOutput): number => accumulator + component.desiredMemory,
      0,
    ),
  }
}

export function calculateNodesDesiredForInterval(interval: Interval): Interval {
  const desiredNodesByCpu = Math.ceil(interval.desiredCpu / interval.nodes.availableCpu)
  const desiredNodesByMemory = Math.ceil(interval.desiredMemory / interval.nodes.availableMemory)
  const desiredNodesByPods = Math.ceil(interval.desiredPods / interval.nodes.maxPods)

  return {
    ...interval,
    desiredNodesByCpu,
    desiredNodesByMemory,
    desiredNodesByPods,
    desiredNodes: withinRange(
      interval.nodes.minNodes,
      interval.nodes.maxNodes,
      Math.max(desiredNodesByCpu, desiredNodesByMemory, desiredNodesByPods),
    ),
  }
}

export function calculateReadyNodes(interval: Interval, index: number, intervals: Interval[]): Interval {
  return {
    ...interval,
    readyNodes:
      index > interval.nodes.scalingIntervals
        ? intervals[index + 1 - interval.nodes.scalingIntervals].desiredNodes
        : interval.nodes.minNodes,
  }
}

export function calculatePendingPods(interval: Interval, index: number, intervals: Interval[]): Interval {
  const intervalComponents = interval.components.map((component: ComponentOutput, cid: number) => {
    let pendingReplica = component.minReplica

    if (index + 1 - component.scalingIntervals > 0) {
      pendingReplica = intervals[index + 1 - component.scalingIntervals].components[cid].desiredReplica
    }

    return {
      ...component,
      pendingReplica,
    }
  })

  return { ...interval, components: intervalComponents }
}

export function calculateReadyPods(interval: Interval): Interval {
  const intervalComponents = interval.components.map((component: ComponentOutput): ComponentOutput => {
    const readyReplica = component.pendingReplica
    return {
      ...component,
      readyReplica,
    }
  })

  let cid = 0
  do {
    intervalComponents[cid].readyReplica = Math.max(
      intervalComponents[cid].readyReplica - 1,
      intervalComponents[cid].minReplica,
    )
    cid = cid === intervalComponents.length - 1 ? 0 : cid + 1
  } while (calculateIfOverSubscribed(intervalComponents, interval.nodes, interval.readyNodes))

  return { ...interval, components: intervalComponents }
}

export function calculateIfOverSubscribed(
  components: ComponentOutput[],
  nodes: ComputeNode,
  readyNodes: number,
): boolean {
  const pendingPodCount = components.reduce(
    (accumulator: number, component: ComponentOutput): number => accumulator + component.readyReplica,
    0,
  )
  if (pendingPodCount > readyNodes * nodes.maxPods) {
    return true
  }
  const pendingCpu = components.reduce(
    (accumulator: number, component: ComponentOutput): number =>
      accumulator + component.readyReplica * component.limitCpu,
    0,
  )

  if (pendingCpu > readyNodes * nodes.availableCpu) {
    return true
  }

  const pendingMemory = components.reduce(
    (accumulator: number, component: ComponentOutput): number =>
      accumulator + component.readyReplica * component.limitMemory,
    0,
  )
  if (pendingMemory > readyNodes * nodes.availableMemory) {
    return true
  }

  return false
}

export function calculateCosts(interval: Interval): Interval {
  return { ...interval, cost: interval.desiredNodes * interval.nodes.cost }
}

export function calculatePenalties(failedRequestPenalty: number) {
  return (interval: Interval): Interval => {
    return {
      ...interval,
      failedRequestPenalty: interval.failedRequests * failedRequestPenalty,
    }
  }
}

export function calculateReadyRequestCapacity(interval: Interval): Interval {
  const intervalComponents = interval.components.map((component) => {
    const cpuCapacity = (component.readyReplica * (component.limitCpu - component.baselineCpu)) / component.requestToCpu

    const memoryCapacity =
      (component.readyReplica * (component.limitMemory - component.baselineMemory)) / component.requestToMemory

    const readyRequestCapacity = Math.floor(Math.min(cpuCapacity, memoryCapacity))
    return { ...component, readyRequestCapacity }
  })
  return {
    ...interval,
    components: intervalComponents,
    readyRequestCapacity: Math.min(...intervalComponents.map((component) => component.readyRequestCapacity)),
  }
}

export function calculateFailedRequests(interval: Interval): Interval {
  const intervalComponents = interval.components.map((component) => {
    return {
      ...component,
      failedRequests: Math.ceil(Math.max(0, interval.requests - component.readyRequestCapacity)),
    }
  })
  return {
    ...interval,
    components: intervalComponents,
    failedRequests: Math.max.apply(
      null,
      intervalComponents.map((component) => component.failedRequests),
    ),
  }
}
