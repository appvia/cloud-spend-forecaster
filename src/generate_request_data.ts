export interface Puller {
  time: number
  requests: number
  desc?: String
}

export interface RequestData {
  time: number
  requests: number
}

export function generate (pullers: Puller[], units: number): RequestData[] {
  const data = []
  pullers.forEach((from: Puller, i: number) => {
    if (i + 1 == pullers.length) return
    const next: Puller = pullers[i + 1]

    const toCreate = (next.time - from.time) / units
    for (let i = 0; i < toCreate; i++) {
      const foo =
        ((next.requests - from.requests) / toCreate) * i + from.requests
      data.push({
        time: i * units + from.time,
        requests: Math.ceil(foo)
      })
    }
  })
  return data
}
