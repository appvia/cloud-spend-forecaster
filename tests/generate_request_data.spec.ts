import { generate } from '../src/generate_request_data'

describe('generate_request_data', () => {
  it('generate with 2 with increasing numbers', () =>
    expect(
      generate(
        [
          { time: 100, requests: 1 },
          { time: 1000, requests: 10000 },
        ],
        10,
      ),
    ).toMatchSnapshot())
  it('generate with 2 with decreasing numbers', () =>
    expect(
      generate(
        [
          { time: 100, requests: 10000 },
          { time: 1000, requests: 1 },
        ],
        10,
      ),
    ).toMatchSnapshot())

  it('generate with increasing and decreasing', () =>
    expect(
      generate(
        [
          { time: 0, requests: 100 },
          { time: 100, requests: 10000 },
          { time: 1000, requests: 1 },
        ],
        10,
      ),
    ).toMatchSnapshot())
})
