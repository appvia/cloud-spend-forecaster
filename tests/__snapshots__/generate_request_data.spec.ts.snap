// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate_request_data generate with 2 with decreasing numbers 1`] = `
Array [
  Object {
    "requests": 10000,
    "time": 100,
  },
  Object {
    "requests": 9889,
    "time": 110,
  },
  Object {
    "requests": 9778,
    "time": 120,
  },
  Object {
    "requests": 9667,
    "time": 130,
  },
  Object {
    "requests": 9556,
    "time": 140,
  },
  Object {
    "requests": 9445,
    "time": 150,
  },
  Object {
    "requests": 9334,
    "time": 160,
  },
  Object {
    "requests": 9223,
    "time": 170,
  },
  Object {
    "requests": 9112,
    "time": 180,
  },
  Object {
    "requests": 9001,
    "time": 190,
  },
  Object {
    "requests": 8889,
    "time": 200,
  },
  Object {
    "requests": 8778,
    "time": 210,
  },
  Object {
    "requests": 8667,
    "time": 220,
  },
  Object {
    "requests": 8556,
    "time": 230,
  },
  Object {
    "requests": 8445,
    "time": 240,
  },
  Object {
    "requests": 8334,
    "time": 250,
  },
  Object {
    "requests": 8223,
    "time": 260,
  },
  Object {
    "requests": 8112,
    "time": 270,
  },
  Object {
    "requests": 8001,
    "time": 280,
  },
  Object {
    "requests": 7890,
    "time": 290,
  },
  Object {
    "requests": 7778,
    "time": 300,
  },
  Object {
    "requests": 7667,
    "time": 310,
  },
  Object {
    "requests": 7556,
    "time": 320,
  },
  Object {
    "requests": 7445,
    "time": 330,
  },
  Object {
    "requests": 7334,
    "time": 340,
  },
  Object {
    "requests": 7223,
    "time": 350,
  },
  Object {
    "requests": 7112,
    "time": 360,
  },
  Object {
    "requests": 7001,
    "time": 370,
  },
  Object {
    "requests": 6890,
    "time": 380,
  },
  Object {
    "requests": 6779,
    "time": 390,
  },
  Object {
    "requests": 6667,
    "time": 400,
  },
  Object {
    "requests": 6556,
    "time": 410,
  },
  Object {
    "requests": 6445,
    "time": 420,
  },
  Object {
    "requests": 6334,
    "time": 430,
  },
  Object {
    "requests": 6223,
    "time": 440,
  },
  Object {
    "requests": 6112,
    "time": 450,
  },
  Object {
    "requests": 6001,
    "time": 460,
  },
  Object {
    "requests": 5890,
    "time": 470,
  },
  Object {
    "requests": 5779,
    "time": 480,
  },
  Object {
    "requests": 5668,
    "time": 490,
  },
  Object {
    "requests": 5556,
    "time": 500,
  },
  Object {
    "requests": 5445,
    "time": 510,
  },
  Object {
    "requests": 5334,
    "time": 520,
  },
  Object {
    "requests": 5223,
    "time": 530,
  },
  Object {
    "requests": 5112,
    "time": 540,
  },
  Object {
    "requests": 5001,
    "time": 550,
  },
  Object {
    "requests": 4890,
    "time": 560,
  },
  Object {
    "requests": 4779,
    "time": 570,
  },
  Object {
    "requests": 4668,
    "time": 580,
  },
  Object {
    "requests": 4557,
    "time": 590,
  },
  Object {
    "requests": 4445,
    "time": 600,
  },
  Object {
    "requests": 4334,
    "time": 610,
  },
  Object {
    "requests": 4223,
    "time": 620,
  },
  Object {
    "requests": 4112,
    "time": 630,
  },
  Object {
    "requests": 4001,
    "time": 640,
  },
  Object {
    "requests": 3890,
    "time": 650,
  },
  Object {
    "requests": 3779,
    "time": 660,
  },
  Object {
    "requests": 3668,
    "time": 670,
  },
  Object {
    "requests": 3557,
    "time": 680,
  },
  Object {
    "requests": 3446,
    "time": 690,
  },
  Object {
    "requests": 3334,
    "time": 700,
  },
  Object {
    "requests": 3223,
    "time": 710,
  },
  Object {
    "requests": 3112,
    "time": 720,
  },
  Object {
    "requests": 3001,
    "time": 730,
  },
  Object {
    "requests": 2890,
    "time": 740,
  },
  Object {
    "requests": 2779,
    "time": 750,
  },
  Object {
    "requests": 2668,
    "time": 760,
  },
  Object {
    "requests": 2557,
    "time": 770,
  },
  Object {
    "requests": 2446,
    "time": 780,
  },
  Object {
    "requests": 2335,
    "time": 790,
  },
  Object {
    "requests": 2223,
    "time": 800,
  },
  Object {
    "requests": 2112,
    "time": 810,
  },
  Object {
    "requests": 2001,
    "time": 820,
  },
  Object {
    "requests": 1890,
    "time": 830,
  },
  Object {
    "requests": 1779,
    "time": 840,
  },
  Object {
    "requests": 1668,
    "time": 850,
  },
  Object {
    "requests": 1557,
    "time": 860,
  },
  Object {
    "requests": 1446,
    "time": 870,
  },
  Object {
    "requests": 1335,
    "time": 880,
  },
  Object {
    "requests": 1224,
    "time": 890,
  },
  Object {
    "requests": 1112,
    "time": 900,
  },
  Object {
    "requests": 1001,
    "time": 910,
  },
  Object {
    "requests": 890,
    "time": 920,
  },
  Object {
    "requests": 779,
    "time": 930,
  },
  Object {
    "requests": 668,
    "time": 940,
  },
  Object {
    "requests": 557,
    "time": 950,
  },
  Object {
    "requests": 446,
    "time": 960,
  },
  Object {
    "requests": 335,
    "time": 970,
  },
  Object {
    "requests": 224,
    "time": 980,
  },
  Object {
    "requests": 113,
    "time": 990,
  },
]
`;

exports[`generate_request_data generate with 2 with increasing numbers 1`] = `
Array [
  Object {
    "requests": 1,
    "time": 100,
  },
  Object {
    "requests": 113,
    "time": 110,
  },
  Object {
    "requests": 224,
    "time": 120,
  },
  Object {
    "requests": 335,
    "time": 130,
  },
  Object {
    "requests": 446,
    "time": 140,
  },
  Object {
    "requests": 557,
    "time": 150,
  },
  Object {
    "requests": 668,
    "time": 160,
  },
  Object {
    "requests": 779,
    "time": 170,
  },
  Object {
    "requests": 890,
    "time": 180,
  },
  Object {
    "requests": 1001,
    "time": 190,
  },
  Object {
    "requests": 1112,
    "time": 200,
  },
  Object {
    "requests": 1224,
    "time": 210,
  },
  Object {
    "requests": 1335,
    "time": 220,
  },
  Object {
    "requests": 1446,
    "time": 230,
  },
  Object {
    "requests": 1557,
    "time": 240,
  },
  Object {
    "requests": 1668,
    "time": 250,
  },
  Object {
    "requests": 1779,
    "time": 260,
  },
  Object {
    "requests": 1890,
    "time": 270,
  },
  Object {
    "requests": 2001,
    "time": 280,
  },
  Object {
    "requests": 2112,
    "time": 290,
  },
  Object {
    "requests": 2223,
    "time": 300,
  },
  Object {
    "requests": 2335,
    "time": 310,
  },
  Object {
    "requests": 2446,
    "time": 320,
  },
  Object {
    "requests": 2557,
    "time": 330,
  },
  Object {
    "requests": 2668,
    "time": 340,
  },
  Object {
    "requests": 2779,
    "time": 350,
  },
  Object {
    "requests": 2890,
    "time": 360,
  },
  Object {
    "requests": 3001,
    "time": 370,
  },
  Object {
    "requests": 3112,
    "time": 380,
  },
  Object {
    "requests": 3223,
    "time": 390,
  },
  Object {
    "requests": 3334,
    "time": 400,
  },
  Object {
    "requests": 3446,
    "time": 410,
  },
  Object {
    "requests": 3557,
    "time": 420,
  },
  Object {
    "requests": 3668,
    "time": 430,
  },
  Object {
    "requests": 3779,
    "time": 440,
  },
  Object {
    "requests": 3890,
    "time": 450,
  },
  Object {
    "requests": 4001,
    "time": 460,
  },
  Object {
    "requests": 4112,
    "time": 470,
  },
  Object {
    "requests": 4223,
    "time": 480,
  },
  Object {
    "requests": 4334,
    "time": 490,
  },
  Object {
    "requests": 4445,
    "time": 500,
  },
  Object {
    "requests": 4557,
    "time": 510,
  },
  Object {
    "requests": 4668,
    "time": 520,
  },
  Object {
    "requests": 4779,
    "time": 530,
  },
  Object {
    "requests": 4890,
    "time": 540,
  },
  Object {
    "requests": 5001,
    "time": 550,
  },
  Object {
    "requests": 5112,
    "time": 560,
  },
  Object {
    "requests": 5223,
    "time": 570,
  },
  Object {
    "requests": 5334,
    "time": 580,
  },
  Object {
    "requests": 5445,
    "time": 590,
  },
  Object {
    "requests": 5556,
    "time": 600,
  },
  Object {
    "requests": 5668,
    "time": 610,
  },
  Object {
    "requests": 5779,
    "time": 620,
  },
  Object {
    "requests": 5890,
    "time": 630,
  },
  Object {
    "requests": 6001,
    "time": 640,
  },
  Object {
    "requests": 6112,
    "time": 650,
  },
  Object {
    "requests": 6223,
    "time": 660,
  },
  Object {
    "requests": 6334,
    "time": 670,
  },
  Object {
    "requests": 6445,
    "time": 680,
  },
  Object {
    "requests": 6556,
    "time": 690,
  },
  Object {
    "requests": 6667,
    "time": 700,
  },
  Object {
    "requests": 6779,
    "time": 710,
  },
  Object {
    "requests": 6890,
    "time": 720,
  },
  Object {
    "requests": 7001,
    "time": 730,
  },
  Object {
    "requests": 7112,
    "time": 740,
  },
  Object {
    "requests": 7223,
    "time": 750,
  },
  Object {
    "requests": 7334,
    "time": 760,
  },
  Object {
    "requests": 7445,
    "time": 770,
  },
  Object {
    "requests": 7556,
    "time": 780,
  },
  Object {
    "requests": 7667,
    "time": 790,
  },
  Object {
    "requests": 7778,
    "time": 800,
  },
  Object {
    "requests": 7890,
    "time": 810,
  },
  Object {
    "requests": 8001,
    "time": 820,
  },
  Object {
    "requests": 8112,
    "time": 830,
  },
  Object {
    "requests": 8223,
    "time": 840,
  },
  Object {
    "requests": 8334,
    "time": 850,
  },
  Object {
    "requests": 8445,
    "time": 860,
  },
  Object {
    "requests": 8556,
    "time": 870,
  },
  Object {
    "requests": 8667,
    "time": 880,
  },
  Object {
    "requests": 8778,
    "time": 890,
  },
  Object {
    "requests": 8889,
    "time": 900,
  },
  Object {
    "requests": 9001,
    "time": 910,
  },
  Object {
    "requests": 9112,
    "time": 920,
  },
  Object {
    "requests": 9223,
    "time": 930,
  },
  Object {
    "requests": 9334,
    "time": 940,
  },
  Object {
    "requests": 9445,
    "time": 950,
  },
  Object {
    "requests": 9556,
    "time": 960,
  },
  Object {
    "requests": 9667,
    "time": 970,
  },
  Object {
    "requests": 9778,
    "time": 980,
  },
  Object {
    "requests": 9889,
    "time": 990,
  },
]
`;

exports[`generate_request_data generate with increasing and decreasing 1`] = `
Array [
  Object {
    "requests": 100,
    "time": 0,
  },
  Object {
    "requests": 1090,
    "time": 10,
  },
  Object {
    "requests": 2080,
    "time": 20,
  },
  Object {
    "requests": 3070,
    "time": 30,
  },
  Object {
    "requests": 4060,
    "time": 40,
  },
  Object {
    "requests": 5050,
    "time": 50,
  },
  Object {
    "requests": 6040,
    "time": 60,
  },
  Object {
    "requests": 7030,
    "time": 70,
  },
  Object {
    "requests": 8020,
    "time": 80,
  },
  Object {
    "requests": 9010,
    "time": 90,
  },
  Object {
    "requests": 10000,
    "time": 100,
  },
  Object {
    "requests": 9889,
    "time": 110,
  },
  Object {
    "requests": 9778,
    "time": 120,
  },
  Object {
    "requests": 9667,
    "time": 130,
  },
  Object {
    "requests": 9556,
    "time": 140,
  },
  Object {
    "requests": 9445,
    "time": 150,
  },
  Object {
    "requests": 9334,
    "time": 160,
  },
  Object {
    "requests": 9223,
    "time": 170,
  },
  Object {
    "requests": 9112,
    "time": 180,
  },
  Object {
    "requests": 9001,
    "time": 190,
  },
  Object {
    "requests": 8889,
    "time": 200,
  },
  Object {
    "requests": 8778,
    "time": 210,
  },
  Object {
    "requests": 8667,
    "time": 220,
  },
  Object {
    "requests": 8556,
    "time": 230,
  },
  Object {
    "requests": 8445,
    "time": 240,
  },
  Object {
    "requests": 8334,
    "time": 250,
  },
  Object {
    "requests": 8223,
    "time": 260,
  },
  Object {
    "requests": 8112,
    "time": 270,
  },
  Object {
    "requests": 8001,
    "time": 280,
  },
  Object {
    "requests": 7890,
    "time": 290,
  },
  Object {
    "requests": 7778,
    "time": 300,
  },
  Object {
    "requests": 7667,
    "time": 310,
  },
  Object {
    "requests": 7556,
    "time": 320,
  },
  Object {
    "requests": 7445,
    "time": 330,
  },
  Object {
    "requests": 7334,
    "time": 340,
  },
  Object {
    "requests": 7223,
    "time": 350,
  },
  Object {
    "requests": 7112,
    "time": 360,
  },
  Object {
    "requests": 7001,
    "time": 370,
  },
  Object {
    "requests": 6890,
    "time": 380,
  },
  Object {
    "requests": 6779,
    "time": 390,
  },
  Object {
    "requests": 6667,
    "time": 400,
  },
  Object {
    "requests": 6556,
    "time": 410,
  },
  Object {
    "requests": 6445,
    "time": 420,
  },
  Object {
    "requests": 6334,
    "time": 430,
  },
  Object {
    "requests": 6223,
    "time": 440,
  },
  Object {
    "requests": 6112,
    "time": 450,
  },
  Object {
    "requests": 6001,
    "time": 460,
  },
  Object {
    "requests": 5890,
    "time": 470,
  },
  Object {
    "requests": 5779,
    "time": 480,
  },
  Object {
    "requests": 5668,
    "time": 490,
  },
  Object {
    "requests": 5556,
    "time": 500,
  },
  Object {
    "requests": 5445,
    "time": 510,
  },
  Object {
    "requests": 5334,
    "time": 520,
  },
  Object {
    "requests": 5223,
    "time": 530,
  },
  Object {
    "requests": 5112,
    "time": 540,
  },
  Object {
    "requests": 5001,
    "time": 550,
  },
  Object {
    "requests": 4890,
    "time": 560,
  },
  Object {
    "requests": 4779,
    "time": 570,
  },
  Object {
    "requests": 4668,
    "time": 580,
  },
  Object {
    "requests": 4557,
    "time": 590,
  },
  Object {
    "requests": 4445,
    "time": 600,
  },
  Object {
    "requests": 4334,
    "time": 610,
  },
  Object {
    "requests": 4223,
    "time": 620,
  },
  Object {
    "requests": 4112,
    "time": 630,
  },
  Object {
    "requests": 4001,
    "time": 640,
  },
  Object {
    "requests": 3890,
    "time": 650,
  },
  Object {
    "requests": 3779,
    "time": 660,
  },
  Object {
    "requests": 3668,
    "time": 670,
  },
  Object {
    "requests": 3557,
    "time": 680,
  },
  Object {
    "requests": 3446,
    "time": 690,
  },
  Object {
    "requests": 3334,
    "time": 700,
  },
  Object {
    "requests": 3223,
    "time": 710,
  },
  Object {
    "requests": 3112,
    "time": 720,
  },
  Object {
    "requests": 3001,
    "time": 730,
  },
  Object {
    "requests": 2890,
    "time": 740,
  },
  Object {
    "requests": 2779,
    "time": 750,
  },
  Object {
    "requests": 2668,
    "time": 760,
  },
  Object {
    "requests": 2557,
    "time": 770,
  },
  Object {
    "requests": 2446,
    "time": 780,
  },
  Object {
    "requests": 2335,
    "time": 790,
  },
  Object {
    "requests": 2223,
    "time": 800,
  },
  Object {
    "requests": 2112,
    "time": 810,
  },
  Object {
    "requests": 2001,
    "time": 820,
  },
  Object {
    "requests": 1890,
    "time": 830,
  },
  Object {
    "requests": 1779,
    "time": 840,
  },
  Object {
    "requests": 1668,
    "time": 850,
  },
  Object {
    "requests": 1557,
    "time": 860,
  },
  Object {
    "requests": 1446,
    "time": 870,
  },
  Object {
    "requests": 1335,
    "time": 880,
  },
  Object {
    "requests": 1224,
    "time": 890,
  },
  Object {
    "requests": 1112,
    "time": 900,
  },
  Object {
    "requests": 1001,
    "time": 910,
  },
  Object {
    "requests": 890,
    "time": 920,
  },
  Object {
    "requests": 779,
    "time": 930,
  },
  Object {
    "requests": 668,
    "time": 940,
  },
  Object {
    "requests": 557,
    "time": 950,
  },
  Object {
    "requests": 446,
    "time": 960,
  },
  Object {
    "requests": 335,
    "time": 970,
  },
  Object {
    "requests": 224,
    "time": 980,
  },
  Object {
    "requests": 113,
    "time": 990,
  },
]
`;
